lib/Dialect/MSFT/MSFTPasses.cpp

```diff
@@ -247,6 +248,176 @@ std::unique_ptr<Pass> createLowerToHWPass() {
 } // namespace msft
 } // namespace circt
 
+namespace {
+struct PartitionPass : public PartitionBase<PartitionPass> {
+  void runOnOperation() override;
+
+  void partition(MSFTModuleOp mod);
+  void partition(DesignPartitionOp part, SmallVectorImpl<Operation *> &users);
+
+private:
+  hw::SymbolCache topLevelSyms;
+};
+} // anonymous namespace
+
+/// Fill a symbol cache with all the top level symbols.
+static void populateSymbolCache(mlir::ModuleOp mod, hw::SymbolCache &cache) {
+  for (Operation &op : mod.getBody()->getOperations()) {
+    StringAttr symName = SymbolTable::getSymbolName(&op);
+    if (!symName)
+      continue;
+    // Add the symbol to the cache.
+    cache.addDefinition(symName, &op);
+  }
+  cache.freeze();
+}
+
+void PartitionPass::runOnOperation() {
+  ModuleOp outerMod = getOperation();
+  ::populateSymbolCache(outerMod, topLevelSyms);
+
+  // TODO: sort modules by number of instantiations so we only have to run one
+  // "bubbling" pass.
+  outerMod.walk([this](MSFTModuleOp mod) { partition(mod); });
+}
+
+void PartitionPass::partition(MSFTModuleOp mod) {
+  DenseMap<SymbolRefAttr, SmallVector<Operation *, 4>> partMembers;
```

Comment:
Nit: for SmallVectors, unless we have a good reason to guess the initial size, I think it is preferred to leave the size off (i.e. `SmallVector<Operation *>`), which should have it default to the largest size that fits neatly in a stack allocation.

Reply:
Actually, `DenseMap` stores these as values for keys not present, so it's probably best to keep them as small as possible. I'll change it to 1.